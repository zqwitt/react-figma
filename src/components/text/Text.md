# Text

Analog of `Text` component at [React Native](https://facebook.github.io/react-native/docs/text), 
[React Sketchapp](http://airbnb.io/react-sketchapp/docs/API.html#text). 
Wrapper for Figma [Text](https://www.figma.com/plugin-docs/api/TextNode/).
`Text` component supports autoloading fonts defined through `fontFamily`, `fontWeight`, `fontStyle` props. 

#### Props

| Prop       | Type     | Default | Note                                              |
| ---------- | -------- | ------- | ------------------------------------------------- |
| `name`     | `String` |         | The name to be displayed in the Figma Layers List |
| `children` | `String` |         | Text content                                      |
| `style`    | [`Style`](/docs/styling.md#type-styles)   |         |                  |
| `onSelectionEnter` | `Function` |  | Selection enter event callback  |
| `onSelectionLeave` | `Function` |  | Selection leave event callback  |

Also, most of the [TextNode](https://www.figma.com/plugin-docs/api/TextNode/) fields supported as props.

#### Example

```javascript
<Text
  style={{
    fontSize: 24,
    fontFamily: 'Helvetica',
    fontWeight: 'bold',
    color: '#01ffae',
  }}
>
  Hello World!
</Text>
```
