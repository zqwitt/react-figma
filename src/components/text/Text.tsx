import * as React from 'react';
import { DefaultShapeProps, StyleOf, TextNodeProps } from '../../types';
import {
    LayoutStyleProperties,
    transformLayoutStyleProperties
} from '../../styleTransformers/transformLayoutStyleProperties';
import {
    TextStyleProperties,
    transformTextStyleProperties
} from '../../styleTransformers/transformTextStyleProperties';
import { useYogaLayout } from '../../hooks/useYogaLayout';
import { transformBlendProperties, BlendStyleProperties } from '../../styleTransformers/transformBlendProperties';
import { YogaStyleProperties } from '../../yoga/YogaStyleProperties';
import { StyleSheet } from '../..';
import { useFontName } from '../../hooks/useFontName';
import { useTextChildren } from '../../hooks/useTextChildren';

export interface TextProps extends TextNodeProps, DefaultShapeProps {
    style?: StyleOf<YogaStyleProperties & LayoutStyleProperties & TextStyleProperties & BlendStyleProperties>;
    children?: string;
}

export const Text: React.FC<TextProps> = props => {
    const nodeRef = React.useRef();

    const style = StyleSheet.flatten(props.style);

    const charactersByChildren = useTextChildren(nodeRef);

    const textProps = {
        ...transformLayoutStyleProperties(style),
        ...transformTextStyleProperties(style),
        ...transformBlendProperties(style),
        ...props,
        characters: charactersByChildren || props.characters
    };
    // @ts-ignore
    const loadedFont = useFontName(textProps.fontName || { family: 'Roboto', style: 'Regular' });
    const yogaProps = useYogaLayout({ nodeRef, ...textProps });
    // @ts-ignore
    return <text {...textProps} {...yogaProps} loadedFont={loadedFont} innerRef={nodeRef} />;
};
