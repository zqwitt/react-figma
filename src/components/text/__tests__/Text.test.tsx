import * as React from 'react';
const renderer = require('react-test-renderer');
import { Text } from '../Text';

describe('<Text />', () => {
    it('Text without props', () => {
        const tree = renderer.create(<Text />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Text with fontFamily', () => {
        const tree = renderer.create(<Text style={{ fontFamily: 'Helvetica' }} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Text with fontFamily, fontWeight', () => {
        const tree = renderer.create(<Text style={{ fontFamily: 'Helvetica', fontWeight: 'bold' }} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Text with fontFamily, fontWeight, fontStyle', () => {
        const tree = renderer
            .create(<Text style={{ fontFamily: 'Helvetica', fontWeight: 'bold', fontStyle: 'italic' }} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
