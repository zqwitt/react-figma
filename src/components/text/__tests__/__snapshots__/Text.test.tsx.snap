// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> Text with fontFamily 1`] = `
<text
  fontName={
    Object {
      "family": "Helvetica",
      "style": "Regular",
    }
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  loadedFont={null}
  style={
    Object {
      "fontFamily": "Helvetica",
    }
  }
/>
`;

exports[`<Text /> Text with fontFamily, fontWeight 1`] = `
<text
  fontName={
    Object {
      "family": "Helvetica",
      "style": "Bold",
    }
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  loadedFont={null}
  style={
    Object {
      "fontFamily": "Helvetica",
      "fontWeight": "bold",
    }
  }
/>
`;

exports[`<Text /> Text with fontFamily, fontWeight, fontStyle 1`] = `
<text
  fontName={
    Object {
      "family": "Helvetica",
      "style": "Bold Italic",
    }
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  loadedFont={null}
  style={
    Object {
      "fontFamily": "Helvetica",
      "fontStyle": "italic",
      "fontWeight": "bold",
    }
  }
/>
`;

exports[`<Text /> Text without props 1`] = `
<text
  innerRef={
    Object {
      "current": undefined,
    }
  }
  loadedFont={null}
/>
`;
